hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java

```diff
@@ -7,584 +7,63 @@
-    // write out any pending records (this can happen when inserts are turned into updates)
-    Iterator<HoodieRecord<T>> newRecordsItr;
-    if (keyToNewRecords instanceof ExternalSpillableMap) {
-      newRecordsItr = ((ExternalSpillableMap) keyToNewRecords).iterator(key -> !writtenRecordKeys.contains(key));
-    } else {
-      newRecordsItr = keyToNewRecords.entrySet().stream()
-          .filter(e -> !writtenRecordKeys.contains(e.getKey()))
-          .map(Map.Entry::getValue)
-          .iterator();
-    }
-    while (newRecordsItr.hasNext()) {
-      HoodieRecord<T> hoodieRecord = newRecordsItr.next();
-      writeInsertRecord(hoodieRecord);
-    }
-  }
-
-  private Schema getNewSchema() {
-    return preserveMetadata ? writeSchemaWithMetaFields : writeSchema;
-  }
-
-  @Override
-  public List<WriteStatus> close() {
-    try {
-      if (isClosed()) {
-        // Handle has already been closed
-        return Collections.emptyList();
-      }
-
-      markClosed();
-      writeIncomingRecords();
-
-      if (keyToNewRecords instanceof Closeable) {
-        ((Closeable) keyToNewRecords).close();
-      }
-
-      keyToNewRecords = null;
-      writtenRecordKeys = null;
-
-      fileWriter.close();
-      fileWriter = null;
-
-      long fileSizeInBytes = storage.getPathInfo(newFilePath).getLength();
-      HoodieWriteStat stat = writeStatus.getStat();
-
-      stat.setTotalWriteBytes(fileSizeInBytes);
-      stat.setFileSizeInBytes(fileSizeInBytes);
-      stat.setNumWrites(recordsWritten);
-      stat.setNumDeletes(recordsDeleted);
-      stat.setNumUpdateWrites(updatedRecordsWritten);
-      stat.setNumInserts(insertRecordsWritten);
-      stat.setTotalWriteErrors(writeStatus.getTotalErrorRecords());
-      RuntimeStats runtimeStats = new RuntimeStats();
-      runtimeStats.setTotalUpsertTime(timer.endTimer());
-      stat.setRuntimeStats(runtimeStats);
-
-      performMergeDataValidationCheck(writeStatus);
-
-      LOG.info("MergeHandle for partitionPath {} fileID {}, took {} ms.", stat.getPartitionPath(),
-          stat.getFileId(), runtimeStats.getTotalUpsertTime());
-
-      return Collections.singletonList(writeStatus);
-    } catch (IOException e) {
-      throw new HoodieUpsertException("Failed to close UpdateHandle", e);
-    }
-  }
-
-  public void performMergeDataValidationCheck(WriteStatus writeStatus) {
-    if (!config.isMergeDataValidationCheckEnabled() || baseFileToMerge == null) {
-      return;
-    }
-
-    long oldNumWrites = 0;
-    try (HoodieFileReader reader = HoodieIOFactory.getIOFactory(hoodieTable.getStorage())
-        .getReaderFactory(this.recordMerger.getRecordType())
-        .getFileReader(config, oldFilePath)) {
-      oldNumWrites = reader.getTotalRecords();
-    } catch (IOException e) {
-      throw new HoodieUpsertException("Failed to check for merge data validation", e);
-    }
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
+  void doMerge() throws IOException;
```

Comment:
So, this is the main method that sub classes need to impl

Reply:
I see based on 
https://github.com/apache/hudi/blob/caa7c6191982c5ae667fa3f28278522c0e4c4897/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/HoodieMergeHelper.java#L133
and 
https://github.com/apache/hudi/blob/caa7c6191982c5ae667fa3f28278522c0e4c4897/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseMergeHelper.java#L54 

write(oldRecord) also needs to be part of the interface. 
we could give it a default empty impl if need be, but it has to be part of the interface for any public apis from merge handles.